File: androidApp/src/main/java/com/mecofarid/summy/screens/game/view/GameScreen.kt
package com.mecofarid.summy.screens.game.view

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.mecofarid.summy.R
import com.mecofarid.summy.common.ui.SystemUiPreview
import com.mecofarid.summy.common.utils.formatTime
import com.mecofarid.summy.features.game.domain.model.GameProgress
import com.mecofarid.summy.features.game.domain.model.Gameplay
import com.mecofarid.summy.features.game.presentation.GameViewModel
import com.mecofarid.summy.features.game.utils.MINUTE_SECOND_PATTERN
import com.mecofarid.summy.resources.AppTheme
import com.mecofarid.summy.resources.Dimens
import com.mecofarid.summy.screens.game.AddendGamePad
import com.mecofarid.summy.screens.game.SumGamePad
import com.mecofarid.summy.screens.game.TargetGamePad
import java.util.*

private const val URL_PRIVACY_POLICY =
    "http://htmlpreview.github.io/?https://github.com/mecoFarid/Summy/blob/main/legal/privacy_policy.html"

@Composable
fun GameScreen(
    gameViewModel: GameViewModel,
    onHandleUrl: (String) -> Unit
) {
    gameViewModel.apply {
        GameScreenContent(
            screenState = screenState.collectAsStateWithLifecycle().value,
            gameplay = gameplay.collectAsStateWithLifecycle().value,
            gameProgress = gameProgress.collectAsStateWithLifecycle().value,
            elapseTime = timeTicker.collectAsStateWithLifecycle().value,
            onAdd = { onAdd(it) },
            onRestartGame = { restartGame() },
            onHandleUrl = onHandleUrl
        )
    }
}

@Composable
fun GameScreenContent(
    screenState: GameViewModel.ScreenState,
    gameplay: Gameplay,
    gameProgress: GameProgress,
    elapseTime: Long,
    onAdd: (Int) -> Unit,
    onRestartGame: () -> Unit,
    onHandleUrl: (String) -> Unit
) {
    Column {
        Column(
            modifier = Modifier
                .weight(1f)
                .padding(Dimens.gu_2)
            ,
            verticalArrangement = Arrangement.SpaceBetween
        ) {
            Box(
                modifier = Modifier.fillMaxWidth()
            ) {
                ProgressIndicator(
                    modifier = Modifier.align(Alignment.CenterStart),
                    text = elapseTime.formatTime(MINUTE_SECOND_PATTERN, Locale.getDefault())
                )
                TargetGamePad(
                    modifier = Modifier.align(Alignment.Center),
                    text = gameplay.target
                )
                ProgressIndicator(
                    modifier = Modifier.align(Alignment.CenterEnd),
                    text = gameProgress.moveCounter.toString()
                )
            }

            SumGamePad(
                modifier = Modifier
                    .align(Alignment.CenterHorizontally),
                text = gameProgress.sum
            )

            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = Dimens.gu_6),
                horizontalArrangement = Arrangement.spacedBy(Dimens.gu_4, Alignment.CenterHorizontally)
            ) {
                gameplay.addends.forEach {
                    AddendGamePad(
                        text = it,
                        onClick = {
                            onAdd(it)
                        }
                    )
                }
            }

            if (screenState.isGameCompleted())
                GameResultDialog(
                    onRestartGame = onRestartGame,
                    gameResult = screenState as GameViewModel.ScreenState.Completed
                )
        }

        Image(
            modifier = Modifier
                .padding(bottom = Dimens.gu_2)
                .clickable {
                    onHandleUrl(URL_PRIVACY_POLICY)
                }
                .clip(
                    RoundedCornerShape(
                        topEnd = Dimens.gu_14,
                        bottomEnd = Dimens.gu_14
                    )
                )
                .background(AppTheme.colorScheme.gameplayPad)
                .padding(Dimens.gu)
                .size(Dimens.gu_3),
            painter = painterResource(id = R.drawable.ic_privacy),
            contentDescription = null
        )
    }
}

@Composable
fun ProgressIndicator(
    modifier: Modifier = Modifier,
    text: String
) {
    Text(
        modifier = modifier,
        text = text,
        style = TextStyle(
            color = AppTheme.colorScheme.gameplayIndicatorText,
            fontSize = AppTheme.dimens.gameplayMediumText
        )
    )
}

@SystemUiPreview
@Composable
fun PreviewGameScreenContent() {
    AppTheme {
        Surface {
            var gameProgress by remember {
                mutableStateOf(GameProgress(459, 78))
            }
            GameScreenContent(
                screenState = GameViewModel.ScreenState.Running,
                gameplay = Gameplay(listOf(1, 24, 3), 105),
                gameProgress = gameProgress,
                elapseTime = 100,
                onAdd = {
                    gameProgress = gameProgress.copy(sum = gameProgress.sum + it)
                },
                onRestartGame = {},
                onHandleUrl = {}
            )
        }
    }
}

File: iosApp/iosApp/Screens/Game/View/GameScreen.swift
import SwiftUI
import shared

struct GameScreen: View {

    @ObservedObject var viewmodel: ViewModel
    
    var body: some View{
        
        GameScreenContent(
            screenState: viewmodel.screenState,
            gameplay: viewmodel.gameplay,
            gameProgress: viewmodel.gameProgress,
            timeTicker: viewmodel.timeTicker,
            onAdd: { addend in
                viewmodel.gameViewmodel.onAdd(addend: addend)
            },
            onRestartGame: {
                viewmodel.gameViewmodel.restartGame()
            }
        )
    }
    
    class ViewModel: ObservableObject{
        
        private(set) var gameViewmodel: GameViewModel
        @Published private(set) var screenState: GameViewModel.ScreenState
        @Published private(set) var gameplay: Gameplay
        @Published private(set) var gameProgress: GameProgress
        @Published private(set) var timeTicker: Int
        
        init(_ gameViewmodel: GameViewModel) {
            self.gameViewmodel = gameViewmodel
            self.screenState = gameViewmodel.screenState.value as! GameViewModel.ScreenState
            self.gameplay = gameViewmodel.gameplay.value as! Gameplay
            self.gameProgress = gameViewmodel.gameProgress.value as! GameProgress
            self.timeTicker = gameViewmodel.timeTicker.value as! Int
            
            self.startObservers()
        }
        
        private func startObservers(){
            gameViewmodel.observe(flow: gameViewmodel.screenState){ state in
                self.screenState = state as! GameViewModel.ScreenState
            }
            gameViewmodel.observe(flow: gameViewmodel.gameplay){ gameplay in
                self.gameplay = gameplay as! Gameplay
            }
            gameViewmodel.observe(flow: gameViewmodel.gameProgress){ gameProgress in
                self.gameProgress = gameProgress as! GameProgress
            }
            gameViewmodel.observe(flow: gameViewmodel.timeTicker){ timeTicker in
                self.timeTicker = timeTicker as! Int
            }
        }
        
        deinit{
            gameViewmodel.clear()
        }
    }
    
}

struct GameScreenContent: View {
    
    private let screenState: GameViewModel.ScreenState
    private let gameplay: Gameplay
    private let gameProgress: GameProgress
    private let timeTicker: Int
    private let onAdd: (Int32) -> Void
    private let onRestartGame: () -> Void
    
    init(
        screenState: GameViewModel.ScreenState,
        gameplay: Gameplay,
        gameProgress: GameProgress,
        timeTicker: Int,
        onAdd: @escaping (Int32) -> Void,
        onRestartGame: @escaping () -> Void
    ) {
        self.screenState = screenState
        self.gameplay = gameplay
        self.gameProgress = gameProgress
        self.timeTicker = timeTicker
        self.onAdd = onAdd
        self.onRestartGame = onRestartGame
    }
    
    var body: some View {
        VStack{
            HStack{
                ProgressIndicator(
                    TimeUtilsKt.formatTime(
                        Int64(timeTicker),
                        pattern: TimeUtilsKt.MINUTE_SECOND_PATTERN,
                        locale: Locale.current
                    )
                ).frame(maxWidth: .infinity, alignment: Alignment.leading)
                
                TargetGamePad(gameplay.target)
                    .frame(maxWidth: .infinity)
                
                ProgressIndicator(String(gameProgress.moveCounter))
                    .frame(maxWidth: .infinity, alignment: Alignment.trailing)
            }
            
            Spacer()
            
            SumGamePad(text: Binding(get: {gameProgress.sum}, set: {_ in}))
            
            Spacer()
            
            HStack{
                ForEach(gameplay.addends, id: \.self){ value in
                    let addend = value.int32Value
                    AddendGamePad(addend){
                        onAdd(addend)
                    }
                    .padding([.leading, .trailing], Dimens.gu_2)
                }
            }
            .padding(.bottom, Dimens.gu_6)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding()
        .background(Colors.background.ignoresSafeArea())
        .dialog(isDialogShowing: screenState is GameViewModel.ScreenStateCompleted){
            GameResultDialog(screenState as! GameViewModel.ScreenStateCompleted){
                onRestartGame()
            }
        }
    }
}

struct ProgressIndicator: View{
    
    private let text: String
    
    init(_ text: String) {
        self.text = text
    }
    
    var body: some View{
        Text(text)
            .foregroundColor(Colors.gameplayIndicatorText)
            .font(Typography.gameplayMediumText)
    }
}

struct ContentView_Previews: PreviewProvider {
	static var previews: some View {
        GameScreenContent(
            screenState: GameViewModel.ScreenStateCompletedFailed(moveCount: 23, elapsedTime: 344),
            gameplay: Gameplay(addends: [KotlinInt(value: 23), KotlinInt(value: 1), KotlinInt(value: 3)], target: 345),
            gameProgress: GameProgress(moveCounter: 23, sum: 145),
            timeTicker: 233,
            onAdd: {_ in },
            onRestartGame: {}
        )
	}
}
